### PR TITLE
Fixes #4

### LINKED ISSUE #4: #to_h serializes all instance methods
Hi,

I'd like to be able to use standard Ruby methods in my `ImmutableStruct`. I'm currently having problems with this one:

```ruby
it "should allow #== method to be overriden" do
  klass = ImmutableStruct.new(:id, :flappy) do
    def ==(other)
      self.id == other.id
    end
  end
  instance = klass.new(id: 1, flappy: 'bird')
  instance.to_h.should == {id: 1, flappy: 'bird'}
end
```

Just because `#to_h` [processes all instance methods](https://github.com/stitchfix/immutable-struct/blob/df5b280520abcac144e46dde4287264a50df7316/lib/immutable-struct.rb#L72), regardless of their intent and it fails when calling methods with arity > 0:

```
1) ImmutableStruct to_h should allow #== method to be overriden
     Failure/Error: def ==(other)
     ArgumentError:
       wrong number of arguments (0 for 1)
     # ./spec/immutable_struct_spec.rb:128:in `=='
     # ./lib/immutable-struct.rb:72:in `block (3 levels) in new'
     # ./lib/immutable-struct.rb:72:in `each'
     # ./lib/immutable-struct.rb:72:in `inject'
     # ./lib/immutable-struct.rb:72:in `block (2 levels) in new'
     # ./spec/immutable_struct_spec.rb:133:in `block (3 levels) in <top (required)>'
```

Moreover, if the user decides to implement `#to_s`, domain-related logic or even a conversion method, these would be serialized as well with unexpected consequences depending on their arity.

IMO this issue raises an interesting question: is it really necessary to serialize all instance methods? In Ruby's [`Struct#to_hash`](http://ruby-doc.org/core-2.2.0/Struct.html#method-i-to_h) only the struct members (ie. those defined in the constructor) are serialized. So I'd expect `ImmutableStruct` to behave in the same way.

If this is OK with you I'd be happy to submit a PR.

Kind regards,
//nando

### PR DESCRIPTION
**Pull request by nandosola**
Well, I guess this PR kinda makes it harder to have default values for attributes. I'm working on something like:

```ruby
Person = ImmutableStruct(:name, :age, location: 'Malaga')
```

which shouldn't be hard to implement.

### CODE REVIEW

File: spec/immutable_struct_spec.rb [hunk 1]
```diff
@@ -103,14 +103,28 @@ def self.from_array(array)
   end
 
   describe "to_h" do
-    it "should include the output of params and block methods in the hash" do
+    it "should not include the output of params and block methods in the hash" do
```
Comment (davetron5000) [inline spec/immutable_struct_spec.rb]: Per my comments on the issue, I'm not in favor of this, at least as the default.  This is a breaking change and we wouldn't be able to merge it without a lot of testing on our end.

File: spec/immutable_struct_spec.rb [hunk 2]
```diff
@@ -103,14 +103,28 @@ def self.from_array(array)
...
         def lawsuit
           'pending'
         end
       end
       instance = klass.new(flappy: 'bird')
-      instance.to_h.should == {flappy: 'bird', lawsuit: 'pending'}
+      instance.to_h.should == {flappy: 'bird'}
     end
+
+    it "allows #== to be overriden" do
```
Comment (davetron5000) [inline spec/immutable_struct_spec.rb]: The issue isn't with the method `==` is it?  It's that nonzero-arity methods are included in the `to_h` implementation.

So, I think this test is mostly fine, but I'd change this line to somethingn like `it "does not include nonzero arity methods in the output"`

Comment (nandosola) [inline spec/immutable_struct_spec.rb]: Yes, it could be changed to that. The issue happened when I subclassed an ImmutableStruct I'm using as an entity class.

**Comment by nandosola** [comment]
Please wait. I've discovered a condition where
```
1) ImmutableStruct serialization to_h can be overriden and return only the attributes defined in the constructor
     Failure/Error: Unable to find matching line from backtrace
     SystemStackError:
       stack level too deep
     # ./lib/immutable-struct.rb:97
```

File: lib/immutable-struct.rb [hunk 3]
```diff
@@ -36,26 +37,46 @@ class ImmutableStruct
...
+  #
+  # Please note that, unless overriden, +ImmutableStruct#to_h+ will serialize all zero-arity methods.
+  #
+  #
   def self.new(*attributes,&block)
+    raise ArgumentError if attributes.empty?
     klass = Class.new do
+      array_attr = lambda { |attr| attr.kind_of?(Array) && attr.size == 1 }
+      boolean_attr = lambda { |attr| String(attr).match(/(^.*)\?$/) }
+      add_reader = lambda { |attr| begin; attr_reader attr; rescue; raise ArgumentError; end }
```
Comment (davetron5000) [inline lib/immutable-struct.rb]: Doesn't seem wise to bury this error—why not just let it raise?

File: lib/immutable-struct.rb [hunk 4]
```diff
@@ -36,26 +37,46 @@ class ImmutableStruct
...
+
       attributes.each do |attribute|
-        if attribute.to_s =~ /(^.*)\?$/
+        case attribute
+        when String
```
Comment (davetron5000) [inline lib/immutable-struct.rb]: Strings work fine—any reason to disallow them?  This is technically a breaking change since strings work now.

**Comment by davetron5000** [comment]
OK, ping back when you have it sorted out—I won't comment on the rest of the code in case you need to make bigger changes to make it work

File: lib/immutable-struct.rb (hunk 3 above)
Comment (nandosola) [inline lib/immutable-struct.rb]: OK You're right. Maybe being too defensive here

File: lib/immutable-struct.rb (hunk 4 above)
Comment (nandosola) [inline lib/immutable-struct.rb]: I see. I just tried to reproduce `Struct`s behavior.  I've just been using this gem for a month, so thanks for the heads-up.

**Comment by nandosola** [comment]
Well, that's pretty much it.  I've made additional changes that take into account your previous feedback. Have a look!

File: lib/immutable-struct.rb [hunk 5]
```diff
@@ -67,9 +85,37 @@ def self.new(*attributes,&block)
...
+
+      derived_methods = imethods - attributes - [:to_h]  #to_h is never a derived method, despite its 0-arity
+
+      ##
+      # :method: attributes_to_h
+      # Return a +Hash+ with only those attributes defined in the class constructor.
+      define_method(:attributes_to_h) do
+        attributes.each_with_object({}) do |attr, hash|
+          ivar = attr.to_s.prepend(?@).to_sym
+          hash[attr] = self.instance_variable_get(ivar)
```
Comment (davetron5000) [inline lib/immutable-struct.rb]: There's no need to turn this into a symbol, you should be able to just do `self.instance_variable_get("@#{attr}")`

File: lib/immutable-struct.rb [hunk 6]
```diff
@@ -67,9 +85,37 @@ def self.new(*attributes,&block)
...
+      # :method: derived_to_h
+      # Return a +Hash+ of the results all zero-arity methods that are not attributes
+      # defined in the class constructor 
+      define_method(:derived_to_h) do
+        derived_methods.each_with_object({}) do |derived_method, hash|
+          hash[derived_method] = self.send(derived_method) if 0 == self.method(derived_method).arity
+        end
+      end
+
+      alias_method :__to_h__, :to_h if method_defined?(:to_h)
```
Comment (davetron5000) [inline lib/immutable-struct.rb]: What's this for?

File: lib/immutable-struct.rb [hunk 7]
```diff
@@ -67,9 +85,37 @@ def self.new(*attributes,&block)
...
+      ##
+      # :method: to_h
+      # Return a +Hash+ with the result of merging +attributes_to_h+ and +derived_to_h+.
+      # This method can also be overriden so that, for example, only +attributes_to_h+ is used.
       define_method(:to_h) do
-        imethods.inject({}){ |hash, method| hash.merge(method.to_sym => self.send(method)) }
+        self.respond_to?(:__to_h__) ? __to_h__ : attributes_to_h.merge(derived_to_h)
```
Comment (davetron5000) [inline lib/immutable-struct.rb]: Why isn't this just `attributes_to_h.merge(derived_to_h)`?

File: spec/immutable_struct_spec.rb [hunk 8]
```diff
@@ -102,15 +112,88 @@ def self.from_array(array)
...
+    end
+
+    describe "derived_to_h" do
+      it "returns a hash of the getters that are not attributes defined in the constructor" do
+        instance = klass.new(flappy: 'bird')
+        instance.derived_to_h.should == {lawsuit: 'pending'}
+      end
+    end
+
+    describe "to_h" do
```
Comment (davetron5000) [inline spec/immutable_struct_spec.rb]: Good test coverage!

File: lib/immutable-struct.rb (hunk 5 above)
Comment (nandosola) [inline lib/immutable-struct.rb]: Nice. thanks!

Comment (nandosola) [inline lib/immutable-struct.rb]: What line do you mean?

Comment (davetron5000) [inline lib/immutable-struct.rb]: I think line 97 can be removed and line 98 can be `hash[attr] = self.instance_variable_get("@#{attr}")`

File: lib/immutable-struct.rb (hunk 6 above)
Comment (nandosola) [inline lib/immutable-struct.rb]: And here? You mean L112?

File: lib/immutable-struct.rb (hunk 5 above)
Comment (nandosola) [inline lib/immutable-struct.rb]: sure! already did.

File: lib/immutable-struct.rb (hunk 6 above)
Comment (davetron5000) [inline lib/immutable-struct.rb]: Yeah, why is this line here/what do we need it for?

File: lib/immutable-struct.rb (hunk 7 above)
Comment (nandosola) [inline lib/immutable-struct.rb]: OK. This is just in case the user decides to override the method. There's already a Spec for this, using:

```ruby
let(:klass_with_overriden_to_h_custom) {
      ImmutableStruct.new(:flappy) do
        def lawsuit
          'pending'
        end
        def to_h
          {:custom => 'hash'}
        end
      end
    }

```

Comment (davetron5000) [inline lib/immutable-struct.rb]: Sorry I'm having a hard time understanding this.  What happens if we remove the `alias` and just have this method merge the two helper methods' hashes? i.e. what goes wrong that the `__to_h__` method is solving?

Comment (nandosola) [inline lib/immutable-struct.rb]: Un saludo,
//nando
https://github.com/nandosola

----
"It is a mistake to think you can solve any major problems just with potatoes"
- Douglas A

Comment (nandosola) [inline lib/immutable-struct.rb]: This `alias_method` handles the case where `#to_h` is overriden by the user:

```ruby
Klass = ImmutableStruct.new(:flappy) do
        def lawsuit
          'pending'
        end
        def to_h
          attributes_to_h  #or another custom logic
        end
      end
```

In the example above, the block is passed to the `ImmutableStruct` constructor and evaluated with `klass.class_exec(&block)` [just before](https://github.com/nandosola/immutable-struct/blob/747e744f9581e07d4265fcdc236dc4df1a1083e0/lib/immutable-struct.rb#L86) defining the `#attributes_to_h`, `#derived_to_h` and `#to_h` instance methods inside `klass.class_exec(imethods, attributes) do |imethods, attributes| ...`

So, if the user has decided to override `#to_h` with some custom conversion (or just calling `#attributes_to_h`) inside an `ImmutableStruct`, then it has to be aliased so that the original `#to_h` method is invoked, thus overriding the default `attributes_to_h.merge(derived_to_h)` .

In my application I'm invoking `#to_h` in different serializable domain classes other than `ImmutableStruct`, so IMHO this would make sense from a duck-typing point of view. TBH, I thought you meant this when we discussed issue #4 a few days ago:

> Then, if you don't wish to have the default implementation of to_h, you have a few options:
> * alias it to attributes_to_h
> * override it completely
> * use attribtues_to_h and derived_to_h however you like to pick and choose

So in my PR the two latter options are possible thanks to the user-defined `#to_h` being aliased to `#__to_h__`, while the first one won't work because `#to_h` cannot be aliased before it's defined, as I stated above.

However, I'm not making a case for anything here. So I'm cool with whatever you decide :-)

**Comment by nandosola** [comment]
Oops! Rubinius build is failing. I'll have a look...

**Comment by nandosola** [comment]
It's most definitely a Rubinius bug. The tests for [`Module#attr_reader`](https://github.com/rubinius/rubinius/blob/eb169f6d44180a610302d395bbce3861db19bd1f/spec/ruby/core/module/attr_reader_spec.rb#L7) seem to forget testing for empty strings/symbols.

While YARV/MRI raises an appropriate exception:
```
irb(main):001:0> class Foo
irb(main):002:1>   attr_reader ''
irb(main):003:1>   end
NameError: invalid attribute name `'
        from (irb):2:in `attr_reader'
        from (irb):2:in `<class:Foo>'
        from (irb):1
        from /Users/nando/.rbenv/versions/2.0.0-p643/bin/irb:12:in `<main>'
```

Rubinius seems to be OK:
```
irb(main):006:0> class Foo
irb(main):007:1>   attr_reader ''                                                                                                                
irb(main):008:1>   end
=> nil
```

Which doesn't makes sense, since Rubinius coerces all attr descriptors to `Symbol`:

```
irb(main):015:0> class Foo                                                                            
irb(main):016:1>   attr_reader 42                                                                                                                
irb(main):017:1>   end
TypeError: Coercion error: 42.to_sym => Symbol failed
```

I have to get some breath. I'll ask politely later in the Rubinius Gitter channel.

**Comment by nandosola** [comment]
ZOMG! This is a bug in JRuby too!
```
irb(main):001:0> class Foo
irb(main):002:1>   attr_reader ''
irb(main):003:1>   end
Java::JavaLang::StringIndexOutOfBoundsException: String index out of range: 0
#...
```

The fun begins.

File: lib/immutable-struct.rb (hunk 7 above)
Comment (davetron5000) [inline lib/immutable-struct.rb]: OK, I think I understand.  It's because if someone overrode `to_h`, that would happen before the constructor is called.

Would it be clearer to do something like:

```ruby
method_name = if defined?(:to_h)
                :to_h_internal
               else
                :to_h
define_method(method_name) do
end
```

I think it's the aliasing that is confusing to me…

**Comment by davetron5000** [comment]
Seems like we shouldn't be creating attributes named for the empty string, so is it a big deal to worry about?

File: lib/immutable-struct.rb (hunk 7 above)
Comment (nandosola) [inline lib/immutable-struct.rb]: I agree with you. We're inside a `#class_exec` and your solution makes the most out of it. Aliasing here is confusing.

**Comment by nandosola** [comment]
Well, all changes applied! WDYT? I guess the JRuby and Rubinius bugs are a completely different yak for me to shave :-)

File: lib/immutable-struct.rb [hunk 9]
```diff
@@ -36,26 +36,47 @@ class ImmutableStruct
...
+    raise ArgumentError if attributes && Array(attributes).empty?
+
     klass = Class.new do
+      array_attr = lambda { |attr| attr.kind_of?(Array) && attr.size == 1 }
+      boolean_attr = lambda { |attr| String(attr).match(/(^.*)\?$/) }
+
       attributes.each do |attribute|
-        if attribute.to_s =~ /(^.*)\?$/
+        case attribute
+          when ''
```
Comment (davetron5000) [inline lib/immutable-struct.rb]: Why do we need this?  I think we can assume no one is passing in the empty string and just let it blow up without a check, right?

Comment (nandosola) [inline lib/immutable-struct.rb]: This is just a defensive hotfix until I fix and report the issue to the Rubinius and JRuby projects. IMHO, it's good that the users of those implementations can reliably use this gem.

If this line is deleted, Rubinius would just silently create an empty `ImmutableStruct` and JRuby would blow up with a naked Java exception. Additionally, this test should be commented-out:

`expect { ImmutableStruct.new('') }.to raise_error(NameError)`

Again, maybe I'm getting ahead of myself here and it'd be fine if I just delete the `when` and comment-out or skip the test.

Comment (davetron5000) [inline lib/immutable-struct.rb]: I guess what I'm saying is that no one should ever expect `ImmutableStruct.new('')` to do anything sensible, so it might be simpler to just ignore this case entirely.

**Comment by davetron5000** [comment]
Sorry I lost track of this.  The new code is not working for boolean and array types, e.g.:

```
2.1.0 :001 > require 'immutable-struct'
 => false 
2.1.0 :002 > Person = ImmutableStruct.new(:name, :minor?, :location, [:aliases])
 => Person 
2.1.0 :003 > x = Person.new(name: "Dave", minor: false, aliases: [ "davetron", "davetron5000"])
 => #<Person:0x007ff5bb1b7548 @name="Dave", @minor=false, @location=nil, @aliases=["davetron", "davetron5000"]> 
2.1.0 :004 > x.to_h
NameError: `@minor?' is not allowed as an instance variable name
	from /Users/davec/Projects/stitchfix/immutable-struct/lib/immutable-struct.rb:91:in `instance_variable_get'
	from /Users/davec/Projects/stitchfix/immutable-struct/lib/immutable-struct.rb:91:in `block (3 levels) in new'
	from /Users/davec/Projects/stitchfix/immutable-struct/lib/immutable-struct.rb:90:in `each'
	from /Users/davec/Projects/stitchfix/immutable-struct/lib/immutable-struct.rb:90:in `each_with_object'
	from /Users/davec/Projects/stitchfix/immutable-struct/lib/immutable-struct.rb:90:in `block (2 levels) in new'
	from /Users/davec/Projects/stitchfix/immutable-struct/lib/immutable-struct.rb:118:in `block (2 levels) in new'
	from (irb):4
	from /Users/davec/.rvm/rubies/ruby-2.1.0/bin/irb:11:in `<main>'
```

**Comment by davetron5000** [comment]
Our test coverage for `to_h` isn't great, so it's easy to see how you missed it.  This is a better test of the current functionality:

https://github.com/stitchfix/immutable-struct/blob/fix-to_h/spec/immutable_struct_spec.rb#L105

**Comment by davetron5000** [comment]
Closing as old.  Please re-open if needed (just some housekeeping :)
